### ISSUE TITLE
Accents and punctuation not in ascii range

### ISSUE THREAD

**Issue by Korcholis** [pndurette/gTTS#25]
Before installing the gtts 1.1.3, I was getting `403 forbidden` errors when using non ASCII characters. After installing this it still doesn't work, but now I get more meaningful errors:
```
root@VFKJ001:~# gtts-cli -l 'es' -o 'test.mp3' '¿ya estás volviendo a casa?'
'ascii' codec can't decode byte 0xc2 in position 0: ordinal not in range(128)
```
It happens that both '¿' and 'á', since are not part of ASCII, are not properly parsed. This problem is in any language non ASCII chars, obviously.

_Also, and probably unrelated to this, after upgrading to 1.1.3 I get this error when there are only ASCII symbols (although the generation is fine in that case):_

```
root@VFKJ001:~# gtts-cli -l 'es' -o 'test.mp3' 'ya estas volviendo a casa?'
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/util/ssl_.py:315: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#snimissingwarning.
  SNIMissingWarning
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/util/ssl_.py:120: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
```

**Comment by Korcholis** [comment]
Funny thing, I rolled one small line back from your 1.1.3 version. In `token.py`, I changed like 25:

````
d = list(bytearray(text.encode('UTF-8')))
````

With:

````
d = list(bytearray(text))
````

Clearly it looks like it's gonna stop working in the early future, but it perfectly worked for me, with both Arabic and Spanish tests being done.

**Comment by pndurette** [comment]
Hmm. The non-ASCII was breaking some new code to generate tokens that would yield 403s. Thanks for this, the encoding etc bits have been moving a lot lately. I'll have to poke and figure out a good solution.

**Comment by Boudewijn26** [comment]
I think the problem is that the text from the argument is incorrectly encoded. I actually wrote a test for this in [gTTS-token](https://github.com/Boudewijn26/gTTS-token/blob/master/gtts_token/tests/test_token.py#L37). Perhaps decoding the args to ASCII and then encoding to UTF-8 might work, although I'm not sure this will work on all platforms.

**Comment by chrisliu529** [comment]
https://github.com/pndurette/gTTS/blob/master/gtts/token.py#L25 doesn't work with Chinese text:
`
d = list(bytearray(text.encode('UTF-8')))
`

A fix can be:
`
d = list(bytearray(unicode(text, 'UTF-8').encode('UTF-8')))
`

A very simple demo:
>>> s='你好'
>>> s.encode('UTF-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 0: ordinal not in range(128)
>>> unicode(s, 'UTF-8').encode('UTF-8')
'\xe4\xbd\xa0\xe5\xa5\xbd'

**Comment by Boudewijn26** [comment]
@chrisliu529 The problem isn't with the code. It is with your demo. Writing `s = '你好'` means that you get an str with an ASCII encoding containing the UTF-8 bytes. Thus encoding it to UTF-8 will fail, because the string is invalid. If you write `s = u'你好'` this will be a unicode (unicode is actually a type) and thus has the UTF-8 encoding and the UTF-8 bytes and thus will succeed.

This is even demonstrated by the error message. ASCII encoding only ranges from 0 - 127, whereas e4 is actually 228.

So Python expects all str to be in ASCII and all unicode to be in UTF-8. You can test this with `type('è')` and `type(u'è')`.

**Comment by chrisliu529** [comment]
@Boudewijn26 
Your reply about string type makes sense.
However, it seems it is not possible to sepcify u'你好' in a CLI interface or in a text file so I reproduced  UnicodeDecodeError in the way above..
I just found my fix above crashed the unit test. I'll provide another fix then issue a pull request later today.

**Comment by Boudewijn26** [comment]
Please make the pull request to [gTTS-token](https://github.com/Boudewijn26/gTTS-token). That project was made to avoid code duplication across other Python projects and I expect that gTTS will depend on it in the near future.

**Comment by chrisliu529** [comment]
Fine. I'll make the pull request to gTTS-token :)

**Comment by Boudewijn26** [comment]
Also I was able to specify `text = u'你好'` in the CLI. [As can be seen in this test.](https://repl.it/BiuQ/0)

**Comment by chrisliu529** [comment]
@Boudewijn26 
That's a good point but actually the CLI interface I mentioned was about  gtts-cli interface like
`gtts-cli -l 'zh' -o 'test.mp3' '你好'`

**Comment by Korcholis** [comment]
Yeah, @Boudewijn26, that's actually the point the whole time. As @chrisliu529 said, you can't tell through command line that the string you're passing is either Unicode or ASCII.

**Comment by chrisliu529** [comment]
I just made a pull request but it didn't work on python 3..
Are we supposed to make gtts work on python 3?

**Comment by Boudewijn26** [comment]
@chrisliu529 Sorry, my bad. I misunderstood.

Yes, Python 3 compatibility is required. I think I have a fix. I will implement it this afternoon.

**Comment by Boudewijn26** [comment]
This should be fixed in gTTS-token 1.1.0.

**Comment by pndurette** [comment]
Closing this, as `gTTS` 1.1.4 now does its token calculations from @Boudewijn26's `gTTS-token`.
